### PR TITLE
PyPI Readiness [minor]

### PR DESCRIPTION
**Pull request by ric-evans**
After a year of using `SetupShop`, we have learned some lessons. This PR address those in preparation for PyPI publication.

TLDR: Running "arbitrary" setup code (`SetupShop`) at package install is not good practice. Let's move that to the pre-publish step (CI/CD). Meanwhile, let's revisit package requirements. After all, does it matter if my client has `thepackage==2.4.1` but my library used `thepackage==2.4.2`? Currently `SetupShop` thinks so :-1:.

##### On every commit
Run `Setup Declaration` & `Generate requirements.txt`. Push changes back to branch
##### On every merge (a commit to master/main)
Run `Setup Declaration`, `Generate requirements.txt`, & `Semantic Release`/`Publish to PYPI`. Push back to main/master.

#### Setup Declaration: `setup.cfg`
PyPA (the `setuptools` developers) recommend using `setup.cfg` for setup boilerplate generation (https://setuptools.pypa.io/en/latest/userguide/declarative_config.html). The most important `setup.cfg` sections for us include `[metadata]`, `[options]`, and `[options.package_data]`. These are needed for PyPI publication. These sections are **nearly 100% generated by CI/CD**. How? 

I Introduce to you the `[wipac:cicd_setup_builder]` section:
```
[wipac:cicd_setup_builder]
pypi_name = wipac-dev-tools
python_min = 3.6
keywords_spaced = python tools utilities
```

Along with executing `setup_builder.py` in CI/CD (GH Action), the following sections are produced and prepended to the existing `setup.cfg` (sections/fields are overridden if they already exist, see in-line comments):
```
[metadata]  # generated by wipac:cicd_setup_builder
name = wipac-dev-tools
version = attr: wipac_dev_tools.__version__
url = https://github.com/WIPACrepo/wipac-dev-tools
author = WIPAC Developers
author_email = developers@icecube.wisc.edu
description = Common, basic, and reusable development tools
long_description = file: README.md, CHANGELOG.md, LICENSE.md
long_description_content_type = text/markdown
keywords =
	python
	tools
	utilities
	WIPAC
	IceCube
license = MIT
classifiers =
	Development Status :: 5 - Production/Stable
	License :: OSI Approved :: MIT License
	Programming Language :: Python :: 3.6
	Programming Language :: Python :: 3.7
	Programming Language :: Python :: 3.8
	Programming Language :: Python :: 3.9
	Programming Language :: Python :: 3.10

[semantic_release]  # generated by wipac:cicd_setup_builder
version_variable = wipac_dev_tools/__init__.py:__version__
upload_to_pypi = True
patch_without_tag = True
commit_parser = semantic_release.history.tag_parser
minor_tag = [minor]
fix_tag = [fix]
branch = main

[options]  # generated by wipac:cicd_setup_builder: 'python_requires', 'packages'
install_requires =
	requests
python_requires = >=3.6, <3.11
packages = find:

[options.package_data]  # generated by wipac:cicd_setup_builder: '*'
* = py.typed
```

`[semantic_release]` is also generated for good measure, this is important for automating PyPI publication.

#### Requirements and `requirements.txt`
As a principle, packages should be developed on the cutting edge. That is, use the latest versions for their requirements. See https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/. To accomplish this, requirements will now go in `setup.cfg`'s `[options]/install_requires`, unpinned (except for rare, known exceptions). **`setup.cfg` is the source of truth for all requirements and `requirements.txt` is its log/snapshot.** `requirements.txt` will be **autogenerated** by parsing `setup.cfg` (using https://github.com/jazzband/pip-tools) in CI/CD (GH Action). `pip-tools` pins these package requirements. This will guarantee we have an easily accessible log of the exact library versions for each of our packages' versions (reproducibility = :smile:). Since this is done in CI/CD, we'll know if something breaks (ie. a requirement's new version isn't backward compatible); see `Future Dependabot`.

Example:
```
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile
#
certifi==2021.10.8
    # via requests
charset-normalizer==2.0.12
    # via requests
idna==3.3
    # via requests
requests==2.27.1
    # via wipac-dev-tools (setup.py)
urllib3==1.26.8
    # via requests
```

Tangent: There is also the `[options.extras_require]` section which will be useful for deploying packages with the WIPAC Telemetry package included/implemented (or excluded).

#### PyPI Publication
Semantic Release takes care of this for us (see `setup.cfg`'s `[semantic_release]/upload_to_pypi` and the `semantic-release.yaml` GH Action).

#### Future Dependabot
Using Github's Dependabot (https://github.blog/2019-05-23-introducing-new-ways-to-keep-your-code-secure/) will catch any security vulnerability, make a PR with the fixes, and thus trigger the CI/CD above.
This will build on / replace @blinkdog's useful `check-requirements.py`

closes https://github.com/WIPACrepo/wipac-dev-tools/issues/20

_Edit: User no longer needs to provide `description`, `url`, and `branch`_

### CODE REVIEW

**Comment by ric-evans** [comment]
This is the new `setup.py`:
```python
from setuptools import setup

setup()

```
That's it :grinning:

**Comment by dsschult** [comment]
Some thoughts:

* `pypi_name`: should we standardize on _ or - as separators?
* `description`: can this be read from github's project description?  would that be a good idea?
* `main_or_master`: a better name might be `branch_name`?  also, can this be inferred from github?

I guess in general I'm looking at what the defaults should be, or how we can find them automatically.

In general this is looking good though.

**Comment by ric-evans** [comment]
> `pypi_name`: should we standardize on _ or - as separators?
- Do you mean for the key or actual value?

> `description`: can this be read from github's project description?  would that be a good idea?
- I suppose it could. Currently, there's no github connection/parsing.

> `main_or_master`: a better name might be `branch_name`?  also, can this be inferred from github?
- I originally had this as `branch_name`, but I figured `main_or_master` gets more to the point that this means "what does your repo call the root branch?" As for inferring from github, see above. This could be done.
 
> I guess in general I'm looking at what the defaults should be, or how we can find them automatically.
- Sure, I'm using this dataclass below. But as for user-friendliness, this could go in a README. Also, maybe in `argparse`'s `-h` printout?
```python
@dataclass
class BuilderSection:
    """Encapsulates the `BUIDLER_SECTION_NAME` section & checks for required/invalid fields."""

    pypi_name: str
    description: str
    url: str
    python_range: str  # python_requires
    main_or_master: str = "main"
    keywords_spaced: str = ""  # comes as "A B C"
```

**Comment by dsschult** [comment]
Other things to think about:
- consider putting the action in a separate repo, and just reference it from there for each project
- consider documentation publishing (this might be another PR?)

**Comment by ric-evans** [comment]
> > > `pypi_name`: should we standardize on _ or - as separators?
> > 
> > 
> > ```
> > * Do you mean for the key or actual value?
> > ```
> 
> I mean the actual value.
- Okay, I think dashes.
> 
> > > `main_or_master`: a better name might be `branch_name`?  also, can this be inferred from github?
> > 
> > 
> > ```
> > * I originally had this as `branch_name`, but I figured `main_or_master` gets more to the point that this means "what does your repo call the root branch?" As for inferring from github, see above. This could be done.
> > ```
> 
> I'm just not sure if github won't change it again. And technically you could pick a different name (I've seen a project that had a different "main" branch).

- Good point. I can go with `branch_name`. `root_branch` is also an option, but perhaps that's vague?

**Comment by ric-evans** [comment]
> consider putting the action in a separate repo, and just reference it from there for each project
- I'm not sure I follow. Do you mean a "symlink" to the separate repo?
> consider documentation publishing (this might be another PR?)
- This will grab `README`, `CHANGELOG`, and `LICENSE` files. That's some documentation, if you mean a `gh-pages`/readthedocs sort of thing, that's out of scope (another PR).

**Comment by ric-evans** [comment]
The only thing left to test is publishing to PyPI (have to wait and see at merge) and the "future dependabot" (which can wait for now)

**Comment by ric-evans** [comment]
Here's an example of CI/CD-driven commits (see history above)
Screenshot from 2022-03-07 10-17-32

File: wipac_dev_tools/setup_builder.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+PythonMinMax = Tuple[Tuple[int, int], Tuple[int, int]]
+
+
+def get_latest_py3_release() -> Tuple[int, int]:
+    """Return the latest python3 release version as tuple."""
+    minor = 10  # start with 3.10
+    while True:
+        url = f"https://docs.python.org/release/3.{minor}.0/"
+        if requests.get(url).status_code == 404:
```
Comment (blinkdog) [inline wipac_dev_tools/setup_builder.py:30]: May want to use `status_code >= 300` in case redirects (3xx) or server errors (5xx) show up; might get stuck in an endless loop if we're looking for a 404 that never comes.

May want to add a cap at minor = 50 and bail out with an error if you reach it.

File: wipac_dev_tools/setup_builder.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        self.url = f"https://github.com/{github_full_repo}"
+
+        _json = requests.get(f"https://api.github.com/repos/{github_full_repo}").json()
+        self.default_branch = cast(str, _json["default_branch"])  # main/master/etc.
+        self.description = cast(str, _json["description"])
+
+
+@dataclass
+class BuilderSection:
+    """Encapsulates the `BUIDLER_SECTION_NAME` section & checks for required/invalid fields."""
```
Comment (blinkdog) [inline wipac_dev_tools/setup_builder.py:48]: `BUILDER_SECTION_NAME` typo again

File: wipac_dev_tools/setup_builder.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    if "py.typed" not in cfg["options.package_data"].get("*", fallback=""):
+        if not cfg["options.package_data"].get("*"):
+            star_data = "py.typed"
+        else:  # append to existing list
+            star_data = f"py.typed, {cfg['options.package_data']['*']}"
+        cfg["options.package_data"]["*"] = star_data
+
+
+def build(setup_cfg: str, github_full_repo: str) -> None:
+    """Build the `setup.cfg` sections according to `BUIDLER_SECTION_NAME`."""
```
Comment (blinkdog) [inline wipac_dev_tools/setup_builder.py:259]: `BUILDER_SECTION_NAME` typo here

File: wipac_dev_tools/setup_builder.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        cfg["options.package_data"] = {}
+
+    # NOTE: 'install_requires' (& 'extras_require') are important and shouldn't be overridden
+    # NOTE: 'entry_points' is to the user's discretion and isn't touched
+
+    _build_out_sections(cfg, os.path.dirname(setup_cfg), github_full_repo)
+
+    # Re-order some sections to the top
+    tops = [
+        BUIDLER_SECTION_NAME,
```
Comment (blinkdog) [inline wipac_dev_tools/setup_builder.py:279]: `BUILDER_SECTION_NAME` typo here

File: setup.cfg [hunk 5]
```diff
@@ -1,36 +1,75 @@
...
+
+[tool:pytest]
+flake8-ignore = E501 E231 E226
+
 [flake8]
 ignore = E226, E231, E501, W503
 
 [coverage:report]
-# Regexes for lines to exclude from consideration
+# regexes for lines to exclude from consideration = None
```
[the remark this answers is not in the document]
Comment (ric-evans) [reply setup.cfg]: Nice catch, I meant to look into that. It's fixed now

File: wipac_dev_tools/setup_builder.py (hunk 1 above)
Comment (ric-evans) [reply wipac_dev_tools/setup_builder.py:30]: Good idea. It'll probably never happen, but this fix is a safer bet.

File: wipac_dev_tools/setup_tools.py [hunk 6]
```diff
@@ -74,6 +74,13 @@ def __init__(
         allow_git_urls: bool = True,
         pinned_packages: Optional[List[str]] = None,
     ):
+        print(  # DEPRECATION WARNING
```
[the remark this answers is not in the document]
Comment (ric-evans) [reply wipac_dev_tools/setup_tools.py:77]: I don't think this is necessary. The current packages out there are always grabbing the latest version of `SetupShop` by design, so killing them is, well, overkill. We have good control over these clients.

File: .github/workflows/setup-and-release.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    steps:
+    - uses: actions/checkout@v2
+      if: ${{ github.ref == 'refs/heads/master' || github.ref == 'refs/heads/main' }}
+      with:
+        fetch-depth: 0
+
+    - name: Python Semantic Release
+      if: ${{ github.ref == 'refs/heads/master' || github.ref == 'refs/heads/main' }}
+      uses: relekang/python-semantic-release@master
```
[the remark this answers is not in the document]
Comment (ric-evans) [reply .github/workflows/setup-and-release.yml:66]: This is their recommended usage (https://python-semantic-release.readthedocs.io/en/latest/automatic-releases/github-actions.html). I do understand your concern.

**Review by dsschult — APPROVED** [review]

**Comment by dsschult** [comment]
Let's see if it works.  Note that as with Dockerhub, you may need to create the project in PyPI first before publishing.

**Comment by ric-evans** [comment]
> Let's see if it works. Note that as with Dockerhub, you may need to create the project in PyPI first before publishing.

It worked, no pre-creation needed!
